core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,173 @@
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Wraps a WatchService and allows paths to be registered.
+ */
+final class PropertiesFileWatcherRegistry {
+    private static final Logger logger = LoggerFactory.getLogger(PropertiesFileWatcherRegistry.class);
+
+    @Nullable
+    private CompletableFuture<Void> future;
+
+    private final Map<String, PropertiesFileWatcherContext> ctxRegistry = new ConcurrentHashMap<>();
+    private final WatchService watchService;
+    private final ExecutorService eventLoop =
+            Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setDaemon(true).build());
+
+    /**
+     * Create a registry using the default file system.
+     */
+    PropertiesFileWatcherRegistry() {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Registers a resourceUrl of the properties file to be watched.
+     * @param resourceUrl url to be watched
+     * @param reloader callback to be called on a file change event
+     */
+    void register(URL resourceUrl, Runnable reloader) {
+        final File file = new File(resourceUrl.getFile());
+        final Path path = file.getParentFile().toPath();
+
+        synchronized (ctxRegistry) {
+            checkArgument(!ctxRegistry.containsKey(resourceUrl.getFile()),
+                          "file is already watched: %s", resourceUrl.getFile());
+            try {
+                final WatchKey key = path.register(watchService,
+                                                   ENTRY_CREATE,
+                                                   ENTRY_MODIFY,
+                                                   ENTRY_DELETE,
+                                                   OVERFLOW);
+                ctxRegistry.put(resourceUrl.getFile(), new PropertiesFileWatcherContext(key, reloader));
+            } catch (IOException e) {
+                throw new IllegalArgumentException("failed to watch file " + resourceUrl.getFile(), e);
+            }
+        }
+
+        startFutureIfPossible();
+    }
+
+    /**
+     * Stops watching a properties file corresponding to the resourceUrl.
+     * @param resourceUrl url to stop watching
+     */
+    void deregister(@Nonnull URL resourceUrl) {
+        final PropertiesFileWatcherContext context = ctxRegistry.remove(resourceUrl.getFile());
+        context.key.cancel();
```

Comment:
Don't we need to `null` check to avoid NPE? (It seems possible to be called more than once )